### PR TITLE
WIP: Refactor to remove global state

### PR DESCRIPTION
**Pull request by devknoll**
Opening PR just to track progress and interest, and promote feedback and discussion. The plan is to factor out Relay's global state.

...ignore the failing tests :smile:

### CODE REVIEW

**Comment by josephsavona** [comment]
@devknoll - great to see you're thinking about this! Moving away from singletons will have to be done incrementally to keep everything working in the interim, but we're happy to provide support and help guide you through these PRs.

I can share an internal doc that I wrote detailing the high-level steps for removing singletons and global state, will follow-up tomorrow.

**Comment by devknoll** [comment]
:+1: That'd be super. My current plan is something along the lines of:

##### PR 1
* Convert singletons to classes with members
* Change internal Relay call sites to use a specified instance (w/ fallback)

##### PR 2
* Introduce `RelayContext` component
* Change singleton accessors to issue a deprecation warning

##### PR 3
* Remove singleton accessors

This should hopefully make the process relatively pain free allowing the changes to flow upstream...

File: src/legacy/store/GraphQLDeferredQueryTracker.js [hunk 1]
```diff
@@ -506,4 +517,7 @@ function getRootCallToIDMap(
   }
   return mapping;
 }
-module.exports = GraphQLDeferredQueryTracker;
+
+module.exports = new GraphQLDeferredQueryTracker(
```
Comment (josephsavona) [inline src/legacy/store/GraphQLDeferredQueryTracker.js]: export the class

File: src/legacy/store/GraphQLQueryRunner.js [hunk 2]
```diff
@@ -311,4 +315,4 @@ function runQueries(
   };
 }
 
-module.exports = GraphQLQueryRunner;
+module.exports = new GraphQLQueryRunner(RelayStoreData.getDefaultInstance());
```
Comment (josephsavona) [inline src/legacy/store/GraphQLQueryRunner.js]: export the class

File: src/store/RelayPendingQueryTracker.js [hunk 3]
```diff
@@ -287,38 +291,45 @@ function hasItems(map: Object): boolean {
...
-
-  // TODO: Use `export type`.
-  PendingFetch,
-
-};
+  }
+}
 
-module.exports = RelayPendingQueryTracker;
+module.exports = new RelayPendingQueryTracker(GraphQLDeferredQueryTracker);
```
Comment (josephsavona) [inline src/store/RelayPendingQueryTracker.js]: same: export the class

File: src/store/RelayStore.js [hunk 4]
```diff
@@ -167,6 +172,7 @@ var RelayStore = {
     var transaction = new RelayMutationTransaction(mutation);
     transaction.commit(callbacks);
   }
-};
+}
 
-module.exports = RelayStore;
+module.exports =
```
Comment (josephsavona) [inline src/store/RelayStore.js]: export the class ;-)

File: src/store/RelayStore.js [hunk 5]
```diff
@@ -167,6 +172,7 @@ var RelayStore = {
     var transaction = new RelayMutationTransaction(mutation);
```
Comment (josephsavona) [inline src/store/RelayStore.js]: for a follow-up: the `RelayStoreData` instance should be passed into the mutation transaction

File: src/store/RelayStore.js [hunk 6]
```diff
@@ -75,7 +75,12 @@ var queuedStore = RelayStoreData.getDefaultInstance().getQueuedStore();
  *  }
  *
  */
-var RelayStore = {
+class RelayStore {
+  _queuedStore: RelayRecordStore;
```
Comment (josephsavona) [inline src/store/RelayStore.js]: `_queryRunner: GraphQLQueryRunner;`

File: src/store/RelayStore.js [hunk 7]
```diff
@@ -75,7 +75,12 @@ var queuedStore = RelayStoreData.getDefaultInstance().getQueuedStore();
...
+
+  constructor(queuedStore: RelayRecordStore) {
+    this._queuedStore = queuedStore;
```
Comment (josephsavona) [inline src/store/RelayStore.js]: `this._queryRunner = new GraphQLQueryRunner(this._storeData);` or similar

File: src/store/RelayStore.js [hunk 8]
```diff
@@ -86,7 +91,7 @@ var RelayStore = {
     callback: ReadyStateChangeCallback
   ): Abortable {
     return GraphQLQueryRunner.run(querySet, callback);
```
Comment (josephsavona) [inline src/store/RelayStore.js]: use the instance of the runner

Comment (josephsavona) [inline src/store/RelayStore.js]: use `this._queryRunner`

**Comment by josephsavona** [comment]
This is a great start. My overall feedback is that any singleton module that is converted to a class should export the class, and all call sites of that module should be updated to either instantiate an instance or be passed one from somewhere. To help understand the implications of each change and make this reviewable, would you mind sending one PR per module?

**Comment by devknoll** [comment]
Ahhh, sorry, wasn't ready yet :wink: I avoided changing exports so that I could ensure that all of the tests were still passing.

**Comment by josephsavona** [comment]
Haha ok :-)

Feel free to keep iterating here while you're exploring, and I'll add comments and feedback. When you're ready let's split into separate PRs.

**Comment by devknoll** [comment]
Can you clarify that you mean you want PRs at e.g. the `RelayMutationTransaction.js` level?

**Comment by josephsavona** [comment]
They don't necessarily have to be exactly one PR per module, but try to think about ways to split it up into incremental changes. Note that `RelayStoreData` is where we are aggregating Relay state, so you can store instances of the new classes on RelayStoreData during the transition. For example if you convert `RelayPendingQueryTracker` to a class and need a place to store the instance (bc you haven't converted QueryRunner yet), you can store the RelayPendingQueryTracker instance in RelayStoreData.

**Comment by josephsavona** [comment]
@devknoll In addition to the previous issue to go with this PR, also take a look at #559 which describes our thoughts on drawing a clear boundary between Relay Core and the Relay/React integration. The proposed `RelayContext` there is a merge of the operations supported by the current `Relay.Store` and `RelayStoreData`.

**Comment by devknoll** [comment]
Closed this to open smaller & more focused PRs.

**Pull request closed without merging**